### PR TITLE
Feature: better selection of base ref branch

### LINKED ISSUE #499: Improve detection of root commit for branches when opening Rebase dashboard
We use a Git Flow-esque branching structure. Consequently, most of my branches are rooted on the `develop` branch; however, the Rebase dashboard always seems to guess I want to use `master` as the base ref. It would seem the logic for detecting the root commit of a branch could use some work.

The `"rebase_default_base_ref"` preference is useful for relieving the pain of this in most cases; however, the branches that _are_ rooted on `master` are then incorrectly compared to `develop` when I open the Rebase dashboard.

### PR DESCRIPTION
**Pull request by asfaltboy**
Before, we used a base ref that was specified in settings, otherwise defaulting to hardcoded "master".

Now, we find the nearest common ancestor branch and use that as our default base, still allowing override in settings.

Fixes #499

-----

The maintainers of this repo require that all pull request submitters agree and adhere to the following:

- [x] I have read and agree to the [contribution guidelines](https://github.com/divmain/GitSavvy/blob/master/CONTRIBUTING.md).
 _(required)_
- [x] All related documentation has been updated to reflect the changes made. _(required)_
- [x] My commit messages are cleaned up and ready to merge. _(required)_

The maintainers of this repository require you to select the semantic version type that
the changes in this pull request represent.  Please select one of the following:
- [ ]  major
- [ ]  minor
- [ ]  patch
- [ ]  documentation only

### CODE REVIEW

**Comment by randy3k** [comment]
Shall we somehow combine [github integrated branch name](https://github.com/divmain/GitSavvy/blob/master/github/git_mixins/remotes.py#L3) with this?

It seems a bit inconsistant that we are using `git config` to set github remote branch while using project settings to set the rebase base.

**Comment by asfaltboy** [comment]
We're not setting the base in project settings, the base is calculated on every render of the view.

The project settings are for getting a default override, which prompted me to suggest #531 , since I believe this should be in (global) default/user settings too.

As for github integrated branch name, I'm not sure there's a better way to set settings persistently in sublime.

**Comment by randy3k** [comment]
The github integrate remote is also [computed](https://github.com/divmain/GitSavvy/blob/33719ef9185fd1fc8d0e77966857dad855af1666/github/git_mixins/remotes.py#L46) every time. `git config` is used to store the default.

**Comment by randy3k** [comment]
Anyway, this PR looks good to me.

**Comment by randy3k** [comment]
Ah....

The `git show-branch` trick does not work if that a branch has merged into the current branch.
The current code will pick the branch `subfeature`, but subfeature has been merge into the current branch (feature).

```
* [feature] Merge branch 'subfeature' into feature
 ! [master] master new commit
  ! [subfeature] subfeature commit
---
 +  [master] master new commit
-   [feature] Merge branch 'subfeature' into feature
* + [subfeature] subfeature commit
*++ [master^] before creating feature branch
```

**Comment by asfaltboy** [comment]
I'm not sure there's a way to intelligently decide for the user which  of the common branches he should rebase on. As far as I know Git doesn't store the original "branching-out" node. In other words, the user could have easily merged "master" into "feature" and it would detect that. If they do merge a sub-feature branch, deleting the merged branch will allow them to rebase on the next detected base.

On the up side, it's still better than what we had (hardcoded "master"), and it can still be overridden in settings if they prefer to keep "base-ref" static.

**Comment by asfaltboy** [comment]
Another option, would be to find the _oldest_ common ancestor, rather than the _nearest_, it would look something like:

```bash
git branch --contains `diff -u <(git rev-list --first-parent develop) \
    <(git rev-list --first-parent master) | sed -ne 's/^ //p' | head -1`
```

But it feels that there's still no "right" answer for all scenarios.

**Comment by asfaltboy** [comment]
@randy3k can you try the command in the above comment on your test case. If it would indicate it's better for you, maybe we can allow both options with a toggle?

**Comment by randy3k** [comment]
The above returns `master` for me, so it stills need to hard code `master`? I have no idea how the above works....

**Comment by asfaltboy** [comment]
Sorry I copy-pasted hastily, I probably needed to update the branches like so: 

```bash
git branch --contains `diff -u <(git rev-list --first-parent feature) \
<(git rev-list --first-parent subfeature) | sed -ne 's/^ //p' | head -1` --no-merged

# or even
git branch --contains `git merge-base feature subfeature` --no-merged
```

Here is how I understand this works, hope it helps:

So, the current base detection (in this PR) works by getting all relative branches, then simply picks the first of those branches, and if it's different than self, returns it.

However, with the above suggested solution, we would not stop at "nearest", rather we would then continue and compare current branch to the potential base branch, looking for the node these trees met (oldest common commit in rev-list). Then check if there is a branch that has this commit and is unmerged, and use that branch as base ref.

**Comment by randy3k** [comment]
The former one returns `master` as expected and the latter one returns nothing...

How about this?

```bash
for s in $(git rev-list --first-parent feature); do 
    branch=`git branch --contains $s --no-merged`
    [ -n "$branch" ] && echo $branch && break
done
```

**Comment by asfaltboy** [comment]
The loops works nicely on my feature branches, but not at all on my master branch, for those it returns a single feature branch (unmerged!), which would be wrong as a base ref.

Correct me if I'm wrong but, the only issue we currently have with the `show-branch` solution above is if we have _many local feature branches_ that share ancestry with current branch, then we may choose those instead of their parents (since children appear before their parents in `show-branch`).

I think I'll add the filter (earliest common no-merge branches) suggested earlier, and see how that works for us.

I still don't think there is a "one size fits all" solution, so we may also want to research a bit more of how others do it.

**Comment by randy3k** [comment]
Sounds a good idea. It would be also handy to preselect the corresponding branch in [GsRebaseOnTopOfCommand](https://github.com/divmain/GitSavvy/blob/master/core/interfaces/rebase.py#L873).

**Comment by asfaltboy** [comment]
Ok ready for review guys! This was a bit troublesome and I want to avoid adding more functionality in this  PR. And to be clear, this is by no means a perfect solution, there may still be cases where the detected base is clearly not what the users wants.

Here's some stuff we need to address:

1. _Refreshing:_ We currently refresh the base only when building the view from scratch again. We could call `base_ref(reset_ref=True)` in `pre_render`, only when a ref is already set, to make sure we have the right ref for the selected branch. This however, will overwrite any custom base ref, meaning `GsRebaseDefineBaseRefCommand` will not work. To address this, we need to change the view setting this command uses, e.g to `user_defined_base_ref` and have it take precedence over the auto-detected / defined in settings. This will make refreshing base in `pre_render` "safe"
2. The base should be used as the pre-selected branch in `GsRebaseOnTopOfCommand`, and other commands of similar nature.
3. A setting can be added to toggle auto-detection of base, though this can already be done by the `rebase_default_base_ref` project setting.

File: core/interfaces/rebase.py [hunk 1]
```diff
@@ -284,6 +296,95 @@ def base_ref(self):
...
+
+        util.debug.add_to_log('nearest_branch: filter out branches that do not'
+                              ' share branch-out nodes')
+        diff = difflib.Differ()
+        branch_commits = self.git("rev-list", "--first-parent", branch).splitlines()
+        nearest = None
+        max_iterations = 100
+        for relative in relatives:
+            util.debug.add_to_log('nearest_branch: Getting common commits with %s' % relative)
+            relative_commits = self.git("rev-list", "--first-parent", relative).splitlines()
```
Comment (randy3k) [inline core/interfaces/rebase.py]: You may set the max iterations by `self.git("rev-list", "-{}".format(max_iterations), "--first-parent", relative)`

File: core/interfaces/rebase.py [hunk 2]
```diff
@@ -258,13 +262,21 @@ def _get_diverged_outside_rebase(self):
...
 
-        if not base_ref:
+        if not base_ref or reset_ref:
             project_data = sublime.active_window().project_data() or {}
             project_settings = project_data.get('settings', {})
-            base_ref = project_settings.get("rebase_default_base_ref", "master")
+            base_ref = project_settings.get("rebase_default_base_ref")
+
+            if not base_ref:
+                # use remote tracking branch as a sane default
```
Comment (randy3k) [inline core/interfaces/rebase.py]: sane -> same

File: core/interfaces/rebase.py [hunk 3]
```diff
@@ -258,13 +262,21 @@ def _get_diverged_outside_rebase(self):
...
+                remote_branch = self._get_branch_status_components()[3]
```
Comment (randy3k) [inline core/interfaces/rebase.py]: `self.get_active_remote_branch().name_with_remote` may be better

File: core/interfaces/rebase.py [hunk 4]
```diff
@@ -284,6 +296,95 @@ def base_ref(self):
 
         return base_ref
 
+    def nearest_branch(self, branch, default="master"):
```
Comment (divmain) [inline core/interfaces/rebase.py]: It's probably worth breaking this out into a helper method, since there's some complexity here and it isn't strictly view-related.

File: core/interfaces/rebase.py [hunk 5]
```diff
@@ -284,6 +296,95 @@ def base_ref(self):
...
+        Solution snagged from:
+        http://stackoverflow.com/a/17843908/484127
+        http://stackoverflow.com/questions/1527234
+        """
+        try:
+            branch_tree = self.git("show-branch", "--no-color").splitlines()
+        except GitSavvyError:
+            return default
+
+        util.debug.add_to_log('nearest_branch: found %s show-branch results' %
```
Comment (divmain) [inline core/interfaces/rebase.py]: Nit: Prefer `"... {}".format(val)` over `"... %s" % val`for consistency.

File: core/interfaces/rebase.py [hunk 6]
```diff
@@ -284,6 +296,95 @@ def base_ref(self):
...
+                continue
+            branch_name = match.groups()[0]
+            if branch_name != branch and branch_name not in relatives:
+                relatives.append(branch_name)
+
+        util.debug.add_to_log('nearest_branch: found %s relatives: %s' %
+                              (len(relatives), relatives))
+
+        if not relatives:
+            util.debug.add_to_log('No relatives found. Possibly on a root branch!')
```
Comment (divmain) [inline core/interfaces/rebase.py]: I like this pattern of verbose debug logging!

File: core/interfaces/rebase.py [hunk 7]
```diff
@@ -284,6 +296,95 @@ def base_ref(self):
...
+                              ' share branch-out nodes')
+        diff = difflib.Differ()
+        branch_commits = self.git("rev-list", "--first-parent", branch).splitlines()
+        nearest = None
+        max_iterations = 100
+        for relative in relatives:
+            util.debug.add_to_log('nearest_branch: Getting common commits with %s' % relative)
+            relative_commits = self.git("rev-list", "--first-parent", relative).splitlines()
+            common = None
+            for i, l in enumerate(diff.compare(branch_commits, relative_commits)):
```
Comment (divmain) [inline core/interfaces/rebase.py]: Prefer descriptive variable names here.

File: core/interfaces/rebase.py [hunk 8]
```diff
@@ -284,6 +296,95 @@ def base_ref(self):
...
+                                  '/ detached branch!')
+            return default
+
+        util.debug.add_to_log('nearest_branch: found best candidate %s' % nearest)
+
+        # if same as branch, return default instead
+        if branch == nearest:
+            return default
+        return nearest
+
```
Comment (divmain) [inline core/interfaces/rebase.py]: This is a huge function.  Would be great if it could be decomposed somehow, but its okay if that would add extra complexity or damage readability.

**Comment by divmain** [comment]
Oh, and last comment.  Would like to make sure this doesn't adversely effect performance too much.  GitSavvy can already be slow, particularly on Windows machines without SSD.  I'd like to avoid making it worse if possible.  I have an older machine that I pull out for things like this, if you would like me to check this.

File: core/interfaces/rebase.py (hunk 4 above)
Comment (asfaltboy) [inline core/interfaces/rebase.py]: Do you mean into it's own helper class? Initially I had this in `BranchesMixin` since it's basically a branch detection mechanism, then I moved it here since felt it's very much related to rebase interface.

But, perhaps it's better if we move this into it's own mixin module: `git_mixins.rebase` to potentially be used by other (rebase) commands. How do you feel about this?

**Comment by asfaltboy** [comment]
@divmain It does add a tiny bit of overhead when opening rebase dashboard, but it's still surprisingly fast on my machine. I would appreciate if we could run this on an old machine! We'd also need a code base with branches that have diverged significantly, to test further.

Other than that, I'd have to use this a bit more to test out more edge cases.

**Comment by randy3k** [comment]
Also related to #369

#369 suggests using remote tracking branch as the base. Perhaps we can use it as a fallback instead of `master`.

**Comment by asfaltboy** [comment]
It's already the default passed by the interface to the helper method. I.e `remote_branch or "master"`

File: core/interfaces/rebase.py (hunk 4 above)
Comment (divmain) [inline core/interfaces/rebase.py]: Yep, moving to `git_mixins.rebase` sounds exactly right.

**Comment by divmain** [comment]
@asfaltboy, let me know when all review comments have been addressed and I'll pull this down onto an old laptop :)

**Comment by stoivo** [comment]
@randy3k  still looking at this?

**Comment by randy3k** [comment]
Hmmm, I could work on this if @asfaltboy agrees.

**Comment by stoivo** [comment]
Ops, i though  I saw you opening this. 
Would be cool if we could close some of the long lasting PR

**Comment by asfaltboy** [comment]
I keep postponing my duties, too much work during the week and weekends are busy. I'll try to address the concerns when able

**Comment by randy3k** [comment]
No problem. Everyone has a real life.

File: core/interfaces/rebase.py [hunk 9]
```diff
@@ -301,13 +302,21 @@ def _get_diverged_outside_rebase(self):
...
-        if not base_ref:
+        if not base_ref or reset_ref:
             project_data = sublime.active_window().project_data() or {}
             project_settings = project_data.get('settings', {})
-            base_ref = project_settings.get("rebase_default_base_ref", "master")
+            base_ref = project_settings.get("rebase_default_base_ref")
+
+            if not base_ref:
+                # use remote tracking branch as a sane default
+                remote_branch = self.get_active_remote_branch().name_with_remote
```
Comment (randy3k) [inline core/interfaces/rebase.py]: `self.get_active_remote_branch()` could be `None` if there is no remote tracking, you probably only need `get_upstream_for_active_branch()`.

**Comment by randy3k** [comment]
It would be more convinenent if `GsRebaseDefineBaseRefCommand` and `GsRebaseOnTopOfCommand` will preselect the chosen `base_ref`.

File: core/interfaces/rebase.py [hunk 10]
```diff
@@ -301,13 +302,21 @@ def _get_diverged_outside_rebase(self):
...
-    def base_ref(self):
+    def base_ref(self, reset_ref=False):
         base_ref = self.view.settings().get("git_savvy.rebase.base_ref")
 
-        if not base_ref:
+        if not base_ref or reset_ref:
             project_data = sublime.active_window().project_data() or {}
             project_settings = project_data.get('settings', {})
-            base_ref = project_settings.get("rebase_default_base_ref", "master")
+            base_ref = project_settings.get("rebase_default_base_ref")
```
Comment (randy3k) [inline core/interfaces/rebase.py]: This is irrelevant to this PR, but I would also like to discuss about this setting `rebase_default_base_ref`. It seems it is the only place we use a project setting while we are using `git config` to set GitHub integration repo and branch.

I just want to make sure there will be a consistent way for storing user settings. So which one shall we use if there is such a need? A: `git config` or B: sublime text project settings.

File: core/interfaces/rebase.py (hunk 9 above)
Comment (randy3k) [reply core/interfaces/rebase.py]: I have fixed it.

File: core/interfaces/rebase.py (hunk 10 above)
Comment (asfaltboy) [reply core/interfaces/rebase.py]: I personally don't think "sublime-specific settings" belong in `git config` ...

My project-settings files are synced to dropbox, and I'd like to avoid defining the override ref again for every new project when I switch machines. Then again, this feature kind of makes this a non-issue.

However, now that you mentioned the Github settings, maybe it could be good to store them in project settings too, unless it makes sense to use these outside of GitSavvy?

**Comment by asfaltboy** [comment]
@randy3k thanks a lot for the review, and the fixes that follow!

Unless @divmain, or anyone with a slow/old/windows machine, wants to profile the rebase interface with this PR, we can go ahead and merge it into develop.

File: core/interfaces/rebase.py (hunk 10 above)
Comment (randy3k) [reply core/interfaces/rebase.py]: The `sublime-project` file might be shared within a team (though it is no likely), the `git config` is not. It might be one of the reasons why historically we were using `git config` rather than project settings.

Though, I agree that those information should not be stored at `git config` neither.
